infer/environment.cc

```diff
@@ -458,6 +460,16 @@ void Environment::updateKnowledge(core::Context ctx, core::LocalVariable local,
                 whoKnows.falsy.mutate().noTypeTests.emplace_back(send->args[0].variable, representedType);
             }
         }
+
+        // `when` against singleton
+        if (auto s = core::cast_type<core::ClassType>(recvType.get())) {
+            // check if s is a singleton. in this case we can learn that
+            // a failed comparison means that type test would also fail
+            if (isSingleton(ctx, s->symbol)) {
```

Comment:
Not featured in this PR, but you might be interested to see the pre-existing definition of isSingleton in Sorbet:

https://github.com/sorbet/sorbet/blob/478a37d0ea1185bc1daf107df992147af4a4bdf7/infer/environment.cc#L334-L337

Complete with a nice little TODO that I get to remove when I add support for arbitrary singletons statically.